desktop_version/src/Music.cpp

```diff
@@ -92,81 +192,327 @@ class SoundTrack
+        
+        if (IsHalted())
+        {
+            SDL_zero(callbacks);
+            callbacks.OnBufferStart = &MusicTrack::refillReserve;
+            callbacks.OnBufferEnd = &MusicTrack::swapBuffers;
+            FAudio_CreateSourceVoice(faudioctx, &musicVoice, &format, 0, 2.0f, &callbacks, NULL, NULL);
+        }
+        else
+        {
+            Pause();
+        }
+        FAudioSourceVoice_FlushSourceBuffers(musicVoice);
+
+        FAudioBuffer faudio_buffer;
+        SDL_memset(&faudio_buffer, 0, sizeof(FAudioBuffer));
+        if (looplength == 0)
         {
-            vlog_error("Mix_PlayMusic: %s", Mix_GetError());
+            faudio_buffer.PlayLength = stb_vorbis_get_samples_float_interleaved(vorbis, channels, (float*)decoded_buf_playing, size/sizeof(float));
+        }
+        else
+        {
+            int samples_read = stb_vorbis_get_samples_float_interleaved(vorbis, channels, (float*)decoded_buf_playing, size/sizeof(float));
+            faudio_buffer.PlayLength = SDL_min(samples_read, (loopbegin+looplength)-sample_pos);
+        }
+        faudio_buffer.AudioBytes = size;
+        faudio_buffer.pAudioData = decoded_buf_playing;
+        faudio_buffer.pContext = this;
+        sample_pos += faudio_buffer.PlayLength;
+        if (FAudioSourceVoice_SubmitSourceBuffer(musicVoice, &faudio_buffer, NULL)) {
+            vlog_error("Unable to queue sound buffer");
             return false;
         }
+        Resume();
         return true;
     }
 
     static void Halt()
     {
-        /* FAudioVoice_Destroy */
-        Mix_HaltMusic();
+        if (!IsHalted())
+        {
+            FAudioSourceVoice_FlushSourceBuffers(musicVoice);
+            FAudioVoice_DestroyVoice(musicVoice);
+            musicVoice = NULL;
+        }
     }
 
     static bool IsHalted()
     {
-        /* return musicVoice == NULL; */
-        return Mix_PausedMusic() == 1;
+        return musicVoice == NULL;
     }
 
     static void Pause()
     {
-        /* FAudioSourceVoice_Pause */
-        Mix_PauseMusic();
+        if (!IsHalted())
+        {
+            FAudioSourceVoice_Stop(musicVoice, 0, FAUDIO_COMMIT_NOW);
+        }
     }
 
     static void Resume()
     {
-        /* FAudioSourceVoice_Resume */
-        Mix_ResumeMusic();
+        if (!IsHalted())
+        {
+            FAudioSourceVoice_Start(musicVoice, 0, FAUDIO_COMMIT_NOW);
+        }
     }
 
     static void SetVolume(int musicVolume)
     {
-        /* FAudioSourceVoice_SetVolume */
-        Mix_VolumeMusic(musicVolume);
+        float adj_vol = (float)musicVolume/VVV_MAX_VOLUME;
+        if (!IsHalted())
+        {
+            FAudioVoice_SetVolume(musicVoice, adj_vol, FAUDIO_COMMIT_NOW);
+        }
     }
 
-private:
-    Mix_Music *m_music;
-};
+    stb_vorbis* vorbis;
+    int channels;
+    Uint32 size;
+    int loopbegin;
+    int looplength;
+    int sample_pos; //stb_vorbis offset not yet functional on pulldata API. TODO Replace when fixed
```

Comment:
What's this TODO about?

Reply:
stb_vorbis has a function to get the current sample position, making tracking it manually unneeded. This does not work yet when using the pulldata API, in contrast to the pushdata API of stb_vorbis. Marked it to know what to remove once this is fixed on stb_vorbis' side. [For reference](https://github.com/FNA-XNA/FAudio/blob/master/src/stb_vorbis.h#L164)